lib/concepts/navigation/navigator_manager.dart

```
@@ -0,0 +1,138 @@
+      // FIXME result is returned before the state is updated, this could be potentially an issue
+      pendingResult.complete(result);
+    }
+  }
+
+  void push(PageData page) {
+    assert(!_disposed,
+        "Tried to use stack manipulation after it was disposed. Are you running the manipulation outside of changeStack method?");
+    _stack.add(page);
+  }
+
+  Future<T?> pushForResult<T>(PageData page) {
+    assert(!_disposed,
+        "Tried to use stack manipulation after it was disposed. Are you running the manipulation outside of changeStack method?");
+    final completer = Completer<T?>();
+    _stack.add(page.copyWith(pendingResult: completer));
+    return completer.future;
+  }
+
+  void replace(PageData page, [dynamic result]) {
+    assert(!_disposed,
+        "Tried to use stack manipulation after it was disposed. Are you running the manipulation outside of changeStack method?");
+    pop(result);
+    push(page);
+  }
+
+  void dispose() {
+    assert(!_disposed,
+        "Tried to use stack manipulation after it was disposed. Are you running the manipulation outside of changeStack method?");
+    _disposed = true;
+  }
+}
+
+abstract class NavigatorManager extends Cubit<xayn.NavigatorState>
+    with UseCaseBlocHelper<xayn.NavigatorState> {
+  final List<PageData<dynamic>> _stack;
+
+  NavigatorManager(List<PageData> initialPages)
+      : _stack = initialPages.isNotEmpty
+            ? initialPages
+            : throw NavigatorException(
+                "At least one initial pages needs to be provided"),
+        super(xayn.NavigatorState(
+            pages: UnmodifiableListView(initialPages.toList(growable: false))));
+
+  @protected
+  T changeStack<T>(T Function(StackManipulation stack) batch) {
+    final manipulation = StackManipulation._(_stack);
+    final lastResult = batch(manipulation);
+    manipulation.dispose();
+    _updateState();
+    return lastResult;
+  }
+
+  @protected
+  Future<T?> pushForResult<T>(PageData page) {
```

💡 not just `push` but also `replace` should have the possibility to return the result. so I guess this method can be merged with `push` and logic duplicated inside `replace` method as well

---

but push and pushForValue are two different mentods, what we need is replace for result, but anyway I am removing all of those methods because they are actually found in the stack manipulation.

---

in the native navigator `push` can return you a result (if I am not mistaken). So why should we create a separate method for that purpose?

---

Because it is a different API. 
Also it can be used to detect errors, like when you pushForResult but you don't receive an error we can throw an exception.

---

> Because it is a different API.

how? it is the same endpoint. but it might return you something and might not. Same way as you have a nullable return type in the current `pushForResult`.

>when you pushForResult but you don't receive an error we can throw an exception.

well, very the same you can do with `Navigator.push`